solr/solrj/src/java/org/apache/solr/common/util/CompressionUtil.java

```diff
@@ -0,0 +1,109 @@
+     * @param data - the bytes to check for compression
+     * @return true if the data is zlib compressed based on the first 2 bytes, false otherwise
+     */
+    public static boolean isCompressedBytes(byte[] data) {
+        if (data == null || data.length < 2) return false;
+        // two characters form the hex value.
+        char[] hexDigits = new char[2];
+        hexDigits[0] = Character.forDigit((data[0] >> 4) & 0xF, 16);
+        hexDigits[1] = Character.forDigit((data[0] & 0xF), 16);
+        if (!Arrays.equals(hexDigits, ZLIB_FIRST_MAGIC_HEX)) return false;
+        hexDigits = new char[2];
+        hexDigits[0] = Character.forDigit((data[1] >> 4) & 0xF, 16);
+        hexDigits[1] = Character.forDigit((data[1] & 0xF), 16);
+        return Arrays.equals(hexDigits, ZLIB_SECOND_MAGIC_HEX);
+    }
+
+    /**
+     * Decompresses zlib compressed bytes, returning the uncompressed data as a byte[]
+     * @param data the input zlib compressed data to decompress
+     * @return the decompressed bytes
+     * @throws DataFormatException - The data is not zlib compressed data
+     */
+    public static byte[] decompressBytes(byte[] data) throws DataFormatException {
+        if (data == null) return null;
+        Inflater inflater = new Inflater();
+        inflater.setInput(data, 0, data.length);
+        // Attempt to get the decompressed size from trailing bytes, this will be present if compressed by Solr
+        int decompressedSize = ByteBuffer.wrap(data, data.length - 8, 4).getInt();
+        int xoredSize = ByteBuffer.wrap(data, data.length - 4, 4).getInt();
+        if ((decompressedSize ^ COMPRESSED_SIZE_MAGIC_NUMBER) != xoredSize) {
+            // Take best guess of decompressed size since it wasn't included in trailing bytes
+            decompressedSize = (int)(1.8*data.length);
+        }
+        byte[] buf = new byte[decompressedSize];
+        int actualDecompressedSize = 0;
+        while (!inflater.finished()) {
+            if (actualDecompressedSize >= buf.length) {
+                buf = Arrays.copyOf(buf, (int)(buf.length*1.5));
+            }
+            actualDecompressedSize += inflater.inflate(buf, actualDecompressedSize, decompressedSize);
+        }
+        return Arrays.copyOf(buf, actualDecompressedSize);
+    }
+
+    /**
+     * Compresses bytes into zlib compressed bytes using java native compression
+     * @param data the input uncompressed data to be compressed
+     * @return zlib compressed bytes
+     * @throws IOException
+     */
+    public static byte[] compressBytes(byte[] data) throws IOException {
+        Deflater compressor = new Deflater(Deflater.BEST_SPEED);
+        compressor.setInput(data);
+        compressor.finish();
+        byte[] buf = new byte[data.length+8];
+        int compressedSize = 0;
+        while (!compressor.finished()) {
+            if (compressedSize >= buf.length) {
+                buf = Arrays.copyOf(buf, (int)(buf.length*1.5));
+            }
```

Comment:
can we move these checks to _after_ the call to `compressor.deflate()`? Assuming we can be 100% certain that we know that `!compressor.finished()` will always be true and `compressedSize >= buf.length` always false the first time we check, which I think is the case?

Reply:
If we move it to after we'll have to have an additional check of compressor.finished() since compressedSize could be equal to buf.length when finished resulting in an additional copy. compressor.finished() is synchronized so I figure this check is lighter even if the first time it isn't needed.

Reply:
Either way's fine really! For comparison, there's some [analogous code in Lucene](https://github.com/apache/lucene/blob/34154736c6ed241d7d9d0c6f4a0e6419936490b7/lucene/core/src/java/org/apache/lucene/codecs/compressing/CompressionMode.java#L284-L294) that I _think_ might address the concern you raise?